### PR TITLE
Performance optimization

### LINKED ISSUE #127: EmojiHandler.addEmojis is horribly slow
Hi! First of all, thanks for this library! I am currently using version 0.5.0-SNAPSHOT (due to issue #103), I don't know if this only happens with this version.

I have seen that `EmojiTextView` and similar are slow when the text is several lines long (I am using a text of about 10000 characters as a test). This blocks the UI thread and is specially noticeable when scrolling a `RecyclerView`.

When running on an emulator (with API 23 to be precise), executing `setText` on an `EmojiTextView` takes about half a second. When doing that in a normal `TextView` the result is immediate (less than 0.001 seconds). I have tried to analyze the behavior of the library and the problematic method is `EmojiHandler.addEmojis`, that seems to parse the full string in a while loop.

I have changed the code in my case to use a regular expression, and the result is amazing: it now takes around 0.002 seconds to execute setText, and the behavior is the same.

Even if you don't take the regexp approach, that method should be checked for optimization.

I attach a Gist of my changed class: https://gist.github.com/Ereza/08282caa4fa444bcbe41aedcc1c820b2

### PR DESCRIPTION
**Pull request by rubengees**
This implements and fixes #127.

### Contents of this Pull Request

Looking up emojis in the `EmojiTextView` and `EmojiEditText` should now be more performant.
I switched to the regex-based apporach, proposed by @Ereza. I noticed bad performance in the `EmojiEditText` when there are many emojis, so I mixed the best of both approaches.
There should be no difference in behaviour.

### Code changes to achieve this

- We create a regex `Pattern` now during the `EmojiManager#install` method.
- I nuked the entire `EmojiTree` and replaced it with a simple `Map`. This is possible, as the regex now acts as our `EmojiTree`. This should be faster, since we now do not need to traverse multiple layers of data structures. It should also be more memory efficient.
-  I moved all finding-logic to the `EmojiManager`. The benefits are:
    - It semantically fits better there imo.
    - We do not need to expose the `Pattern` or other internal logic.
    - We may make the api of the `EmojiManager` public for other use cases. It is able to find emojis in a `CharSequence` and returns the range they are located in. This would also implement most of #128.
- The `EmojiHandler` still has logic for reusing existing `Spans`. This is the difference to the Gist of @Ereza.
- The `EmojiManager#install` method now requires to install an `EmojiProvider` with at least one `EmojiCategory`, having at least one `Emoji`.

I also upgraded `Gradle`, the plugin and the `errorprone` plugin, to make it run in my `Android Studio` 3.0 installation. I hope that's fine. I also had to fix some lint issues which popped up.

### CODE REVIEW

File: emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java [hunk 1]
```diff
@@ -2,83 +2,28 @@
...
+import com.vanniktech.emoji.EmojiManager.EmojiRange;
 import java.util.ArrayList;
 import java.util.List;
 
-import static com.vanniktech.emoji.EmojiHandler.SpanRangeList.SPAN_NOT_FOUND;
-
 final class EmojiHandler {
-  static void addEmojis(final Context context, final Spannable text, final int emojiSize) {
-    final SpanRangeList existingSpanRanges = new SpanRangeList(text);
+  static void replaceWithImages(final Context context, final Spannable text, final int emojiSize) {
```
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java]: Sneaking this name change in, I hope that's fine.

File: build.gradle [hunk 2]
```diff
@@ -20,9 +20,9 @@ buildscript {
         classpath 'com.vanniktech:gradle-code-quality-tools-plugin:0.4.0'
         classpath 'com.vanniktech:gradle-android-junit-jacoco-plugin:0.4.0'
 
-        classpath 'com.android.tools.build:gradle:2.2.2'
+        classpath 'com.android.tools.build:gradle:2.3.2'
         classpath 'com.github.ben-manes:gradle-versions-plugin:0.14.0'
```
Comment (vanniktech) [inline build.gradle]: Then also feel free to bump this one to `0.15.0`

File: build.gradle [hunk 3]
```diff
@@ -20,9 +20,9 @@ buildscript {
         classpath 'com.vanniktech:gradle-code-quality-tools-plugin:0.4.0'
         classpath 'com.vanniktech:gradle-android-junit-jacoco-plugin:0.4.0'
```
Comment (vanniktech) [inline build.gradle]: `0.6.0`

File: emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java [hunk 4]
```diff
@@ -2,83 +2,28 @@
...
-    int index = 0;
-
-    while (index < text.length()) {
-      final int existingSpanEnd = existingSpanRanges.spanEnd(index);
-
-      if (existingSpanEnd == SPAN_NOT_FOUND) {
-        final int nextSpanStart = existingSpanRanges.nextSpanStart(index);
-        final int searchRange = nextSpanStart == SPAN_NOT_FOUND ? text.length() : nextSpanStart;
-        final Emoji found = emojiManager.findEmoji(text.subSequence(index, searchRange));
+    final List<Integer> existingSpans = new ArrayList<>();
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java]: extract `text.getSpans(0, text.length(), EmojiSpan.class)` and then use the size to initialize the List so that it won't increase it's internal limit while adding items to reduce allocations

File: emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java [hunk 5]
```diff
@@ -2,83 +2,28 @@
...
-          index += found.getLength();
-        } else {
-          index++;
-        }
-      } else {
-        index += existingSpanEnd - index;
+    for (final EmojiRange location : emojiManager.findAllEmojis(text)) {
+      if (!existingSpans.contains(location.getStart())) {
+        text.setSpan(new EmojiSpan(context, location.getEmoji().getResource(), emojiSize),
+                location.getStart(), location.getEnd(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java]: you can move all into one line

File: emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java [hunk 6]
```diff
@@ -2,83 +2,28 @@
...
+    }
 
-          index += found.getLength();
-        } else {
-          index++;
-        }
-      } else {
-        index += existingSpanEnd - index;
+    for (final EmojiRange location : emojiManager.findAllEmojis(text)) {
+      if (!existingSpans.contains(location.getStart())) {
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java]: exract `location.getStart()` to save a call

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 7]
```diff
@@ -34,32 +50,98 @@ static EmojiManager getInstance() {
...
 
       //noinspection ForLoopReplaceableByForEach
       for (int j = 0; j < emojis.length; j++) {
-        INSTANCE.emojiTree.add(emojis[j]);
+        final Emoji currentEmoji = emojis[j];
+
+        INSTANCE.emojiMap.put(currentEmoji.getUnicode(), currentEmoji);
+        unicodesForPattern.add(currentEmoji.getUnicode());
+
+        for (final Emoji variant : currentEmoji.getVariants()) {
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: also fori to reduce iterator allocation?

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 8]
```diff
@@ -34,32 +50,98 @@ static EmojiManager getInstance() {
...
     //noinspection ForLoopReplaceableByForEach
     for (int i = 0; i < INSTANCE.categories.length; i++) {
       final Emoji[] emojis = checkNotNull(INSTANCE.categories[i].getEmojis(), "emojies == null");
 
       //noinspection ForLoopReplaceableByForEach
       for (int j = 0; j < emojis.length; j++) {
-        INSTANCE.emojiTree.add(emojis[j]);
+        final Emoji currentEmoji = emojis[j];
+
+        INSTANCE.emojiMap.put(currentEmoji.getUnicode(), currentEmoji);
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: extract `currentEmoji.getUnicode()`

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 9]
```diff
@@ -34,32 +50,98 @@ static EmojiManager getInstance() {
...
+        unicodesForPattern.add(currentEmoji.getUnicode());
+
+        for (final Emoji variant : currentEmoji.getVariants()) {
+          INSTANCE.emojiMap.put(variant.getUnicode(), variant);
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: extract `variant.getUnicode()`

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 10]
```diff
@@ -34,32 +50,98 @@ static EmojiManager getInstance() {
...
+          unicodesForPattern.add(variant.getUnicode());
+        }
       }
     }
+
+    if (unicodesForPattern.isEmpty()) {
+      throw new IllegalArgumentException("Your EmojiProvider must at least have one category with at least one emoji");
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: nit: end the sentence with a dot

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 11]
```diff
@@ -34,32 +50,98 @@ static EmojiManager getInstance() {
...
   }
 
   void verifyInstalled() {
     if (categories == null) {
       throw new IllegalStateException("Please install an EmojiProvider through the EmojiManager.install() method first.");
     }
   }
+
+  static class EmojiRange {
+    private final int start;
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: remove private modifier and the get methods

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 12]
```diff
@@ -34,32 +50,98 @@ static EmojiManager getInstance() {
...
+    private final int end;
+    private final Emoji emoji;
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: remove private modifier and the get methods

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 13]
```diff
@@ -34,32 +50,98 @@ static EmojiManager getInstance() {
...
 
   void verifyInstalled() {
     if (categories == null) {
       throw new IllegalStateException("Please install an EmojiProvider through the EmojiManager.install() method first.");
     }
   }
+
+  static class EmojiRange {
+    private final int start;
+    private final int end;
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: remove private modifier and the get methods

File: emoji/src/main/res/values/library_emoji_strings.xml [hunk 14]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<resources>
```
Comment (rubengees) [inline emoji/src/main/res/values/library_emoji_strings.xml]: Lint complained about unused resources here, failing the build.

File: emoji/src/main/res/values/library_emoji_strings.xml [hunk 15]
```diff
@@ -1,5 +1,5 @@
...
+<resources xmlns:tools="http://schemas.android.com/tools" tools:ignore="UnusedResources">
```
Comment (vanniktech) [inline emoji/src/main/res/values/library_emoji_strings.xml]: Hmm this should be already configured via a lint.xml file

File: emoji/src/main/java/com/vanniktech/emoji/RecentEmojiManager.java [hunk 16]
```diff
@@ -133,7 +133,7 @@ int size() {
       return emojis.size();
     }
 
-    @Override public Iterator<Data> iterator() {
```
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/RecentEmojiManager.java]: Also, lint was complaining about this one.

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java [hunk 17]
```diff
@@ -49,10 +47,26 @@
...
   }
 
+  @Test public void installEmptyCategory() {
+    final EmojiProvider emptyProvider = new EmojiProvider() {
+      @NonNull @Override public EmojiCategory[] getCategories() {
+        return new EmojiCategory[0];
+      }
+    };
+
+    assertThatThrownBy(new ThrowableAssert.ThrowingCallable() {
```
Comment (vanniktech) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: uh nice. TIL 👍

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java [hunk 18]
```diff
@@ -49,10 +47,26 @@
...
+      @Override
+      public void call() throws Throwable {
+        EmojiManager.install(emptyProvider);
+      }
+    }).isInstanceOf(IllegalArgumentException.class);
```
Comment (vanniktech) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: Can you also verify the message here?

File: emoji/src/main/res/values/library_emoji_strings.xml (hunk 15 above)
Comment (rubengees) [inline emoji/src/main/res/values/library_emoji_strings.xml]: Lint complained about unused resources here, failing the build.

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java (hunk 17 above)
Comment (vanniktech) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: also squeeze in another test case where you pass `emptyList()` as the providers

File: emoji/src/main/java/com/vanniktech/emoji/RecentEmojiManager.java [hunk 19]
```diff
@@ -133,7 +133,7 @@ int size() {
...
+    @Override @NonNull public Iterator<Data> iterator() {
```
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/RecentEmojiManager.java]: Also, lint was complaining about this one.

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java [hunk 20]
```diff
@@ -68,4 +82,10 @@
 
     assertThat(EmojiManager.getInstance().findEmoji("")).isNull();
   }
+
+  @Test public void findAllEmojisEmpty() {
```
Comment (vanniktech) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: this only tests the case of `findAllEmojis` when it's empty. It should also test the actual logic if it's not empty

File: emoji/src/main/res/values/library_emoji_strings.xml (hunk 15 above)
Comment (vanniktech) [inline emoji/src/main/res/values/library_emoji_strings.xml]: I think it's a lint bug.

`emoji/lint.xml` contains this:

```xml
<issue id="UnusedResources">
    <ignore regexp="library_emoji_strings.xml"/>
</issue>
```

https://issuetracker.google.com/issues/37981147

Comment (rubengees) [inline emoji/src/main/res/values/library_emoji_strings.xml]: Should we keep this ignore in for the moment, until fixed by the Google team?

File: emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java (hunk 6 above)
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java]: Not necessarily needed anymore, as we access the fields of `EmojiRange` directly now.

File: emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java (hunk 5 above)
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java]: It hurts my eyes :cry:. Required?

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java (hunk 17 above)
Comment (rubengees) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: I guess you meant an empty `EmojiCategory`? If not, could you go in detail what you mean?

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java (hunk 20 above)
Comment (rubengees) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: This is already [entirely covered] by the tests in the `EmojiHandlerTest`. Should i test it nevertheless here?

**Comment by rubengees** [comment]
Most of it done (I have added comments where not).
I also update everything to the respective latest stable versions, while i was at it. If not desired, let me know.

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java [hunk 21]
```diff
@@ -49,10 +47,61 @@
     assertThat(EmojiManager.getInstance().getCategories()).isNotEmpty();
   }
 
+  @Test
+  public void noProviderInstalled() {
```
Comment (rubengees) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: This test is currently failing because `EmojiManager` is a singleton and there is a `EmojiProvider` from a previous test installed.
This led me to the thought of adding a `EmojiManager#destroy` method which would remove the current `EmojiProvider` and clean up all internal data-structures.
This would allow apps to free the memory when they know that the `EmojiManager` is not needed anymore (The user exits the screen with the emoji input for example).

What do you think of this?

File: emoji/src/main/res/values/library_emoji_strings.xml (hunk 15 above)
Comment (vanniktech) [inline emoji/src/main/res/values/library_emoji_strings.xml]: just removei t from `lint.xml` if it works in the xml

File: emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java (hunk 6 above)
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java]: Yes I also realized this then. Even better 👍

File: emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java (hunk 5 above)
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java]: Nah then leave it as is. Or if you already changed it, leave it. Up to you. I've got a big monitor and prefer having it in one but having two lines is also ok

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java [hunk 22]
```diff
@@ -49,10 +47,61 @@
...
+      public void call() throws Throwable {
+        EmojiManager.getInstance().findEmoji("test");
+      }
+    }).isInstanceOf(IllegalStateException.class).hasMessage("Please install an EmojiProvider through the EmojiManager.install() method first.");
+  }
+
+  @Test public void installEmptyProvider() {
+    final EmojiProvider emptyProvider = new EmojiProvider() {
+      @NonNull @Override public EmojiCategory[] getCategories() {
+        return new EmojiCategory[0];
```
Comment (vanniktech) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: Also a test here for having a category but the category not having any Emojis.

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java (hunk 20 above)
Comment (vanniktech) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: Just one simple one so that it does not look like it's not covered.

File: build.gradle [hunk 23]
```diff
@@ -70,16 +70,16 @@ ext {
...
+            assertJ           : 'org.assertj:assertj-core:3.8.0'
     ]
 
     javaVersion = JavaVersion.VERSION_1_7
 }
 
 task wrapper(type: Wrapper) {
-    gradleVersion = '3.3'
+    gradleVersion = '3.5'
     distributionUrl = "https://services.gradle.org/distributions/gradle-$gradleVersion-all.zip"
```
Comment (vanniktech) [inline build.gradle]: when we are anyways changing everything this line can be changed to `distributionType = Wrapper.DistributionType.ALL`

File: build.gradle [hunk 24]
```diff
@@ -7,7 +7,7 @@ codeQualityTools {
     toolVersion = '7.7'
```
Comment (vanniktech) [inline build.gradle]: 7.8 🤓

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 25]
```diff
@@ -13,9 +20,18 @@
  */
 public final class EmojiManager {
   private static final EmojiManager INSTANCE = new EmojiManager();
+  private static final int GUESSED_UNICODE_AMOUNT = 3000;
+  private static final int GUESSED_TOTAL_PATTERN_LENGTH = GUESSED_UNICODE_AMOUNT * 4;
 
-  private final EmojiTree emojiTree = new EmojiTree();
+  private static final Comparator<String> STRING_LENGTH_COMPARATOR = new Comparator<String>() {
+    @Override public int compare(final String first, final String second) {
+      return second.length() - first.length();
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: This can cause issues. Have a look at [this](https://stackoverflow.com/a/2728858/1979703). It's an extreme edge case though, although we should cover it.

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 26]
```diff
@@ -34,32 +50,94 @@ static EmojiManager getInstance() {
...
       //noinspection ForLoopReplaceableByForEach
       for (int j = 0; j < emojis.length; j++) {
-        INSTANCE.emojiTree.add(emojis[j]);
+        final Emoji emoji = emojis[j];
+        final String unicode = emoji.getUnicode();
+        final List<Emoji> variants = emoji.getVariants();
+
+        INSTANCE.emojiMap.put(unicode, emoji);
+        unicodesForPattern.add(unicode);
+
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: `//noinspection ForLoopReplaceableByForEach` so that intellij won't complain

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java [hunk 27]
```diff
@@ -49,10 +47,61 @@
     assertThat(EmojiManager.getInstance().getCategories()).isNotEmpty();
   }
 
+  @Test
```
Comment (vanniktech) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: same line

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java (hunk 21 above)
Comment (vanniktech) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: Let's make it for now package private so it can't be used. I anyways want the install() method to provide some type on which you can do a bunch of strings (like the string queries that I proposed) and this method might get added too.

File: emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java (hunk 17 above)
Comment (vanniktech) [inline emoji/src/test/java/com/vanniktech/emoji/EmojiManagerTest.java]: See above.

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java (hunk 25 above)
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: Oh right! Never thought about it that way.

**Comment by rubengees** [comment]
All done :+1:

**Comment by vanniktech** [comment]
I just wanted to update the branch too. You were a bit faster though.

Once it's done, I'll merge it. Many thanks 👍

**Comment by rubengees** [comment]
Great! What would be missing for a `0.5` release?

**Comment by vanniktech** [comment]
Not much. Just haven't gotten around to cutting a release yet plus my release computer isn't set up right now. Although I'm planning to release 0.5.0 next week.
